python/cudf/cudf/io/parquet.py

```diff
@@ -203,6 +205,48 @@ def read_parquet(
             )
         filepaths_or_buffers.append(tmp_source)
 
+    if filters is not None:
+        # Convert filters to ds.Expression
+        filters = pq._filters_to_expression(filters)
+
+        # Initialize ds.FilesystemDataset
+        dataset = ds.dataset(filepaths_or_buffers, format="parquet")
```

Comment:
Do we know how much / what IO this does relative to the IO that cuIO expects to do? I'm guessing this may be duplicating some of the IO of reading metadata where it would be good to understand the current inefficiencies and see if there's the ability to eliminate the duplicate reads.

Reply:
`ds.dataset` reads the footers of the given Parquet files. I'm not yet sure if it filters at file-level without reading the whole footer. But yes, I/O is duplicated.

This is duplication can only be avoided by making the following changes in libcudf itself.
1. Parse statistics in each row group's metadata [1]
2. Implement evaluation of given predicates against statistics

Currently, on the Python side, we're pretty much just forwarding to pyarrow which does the evaluation of predicates against statistics. So even if we read in the statistics once, we must additionally implement predicate evaluation.

[1] https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/src/io/parquet/parquet.h#L96

Reply:
> `ds.dataset` reads the footers of the given Parquet files. I'm not yet sure if it filters at file-level without reading the whole footer. But yes, I/O is duplicated.

I guess I'm wondering if there's a way to reuse an `mmap` or something along those lines to avoid duplicating unnecessary I/O. That sounds expensive, but we'd need to benchmark to have proof.

Reply:
By default `ds.dataset(..)` itself will only read the metadata of the first file (and assume that the schema of this file can be used for all files in a partitioned dataset), but then the subsequent `get_fragments(filter=...)` call will indeed need to read the metadata of each of the files if it filters on actual data columns.

If you work with partitioned datasets that might have a common `_metadata` file, there is also `ds.parquet_dataset(..)` which only reads the metadata from this single `_metadata` file instead of from all data files.

It is possible to construct a FileSystemDataset from buffers / file-like objects, so I suppose it should also work with mmaps.

(and don't hesitate to open JIRAs about use cases / features that might help using this in cudf)

Reply:
@kkraus14 @jorisvandenbossche Thank you for your comments on this.

Given the benchmarks [below](https://github.com/rapidsai/cudf/pull/5843#issuecomment-673566456), do you think avoiding reading metadata twice is important for this PR? If my understanding is correct, I'm not sure how it could be avoided without passing around metadata (or a pointer to it). The most reasonable way to avoid duplicate reading that I see is implementing it directly in libcudf in a future PR.

Reply:
cc @vuule @devavret for this conversation as well

If we're only reading the metadata on the Python side and then never reading it on the libcudf side then I think we're fine. If the metadata is being read again on the libcudf side it would be ideal to take some kind of path like:
1. Perform metadata IO from Python side via PyArrow and get a host buffer
2. Take 1 and do row group analysis in Python side
3. Take 1 and 2 and perform actual reading / parsing of data via libcudf.

The key is in step 3 would be to reuse an existing host buffer as opposed to redoing IO.

Reply:
So I know that libcudf does read metadata stored in the footers of Parquet files. This would require changing the API to `read_parquet` to optionally accept a host buffer containing metadata for some input files. I'm not aware of any use case for this `metadata` parameter other than our use case. @kkraus14 Is that API change okay?

@jorisvandenbossche Creating a `Dataset` with `dataset` and then filtering by passing predicate into `get_fragments` and `split_by_row_groups` is the only way of determining row groups that pass a given predicate with pyarrow 1.0 - right? Is there actually a way to get metadata that was read in in the process of calling `get_fragments` and `split_by_row_groups` in the way that this PR does?

Reply:
> So I know that libcudf does read metadata stored in the footers of Parquet files. This would require changing the API to `read_parquet` to optionally accept a host buffer containing metadata for some input files. I'm not aware of any use case for this `metadata` parameter other than our use case. @kkraus14 Is that API change okay?

I would 100% defer to @vuule and @devavret's judgements in this area as they're the experts there. I'm just trying to come up with ideas to avoid duplicating IO since it's typically very expensive, especially when you're hitting cloud / network filesystems.

Reply:
I will say that this might be a reason to at least eventually do filtering directly in libcudf.

Reply:
I have various questions
1. Will the data being passed to libcudf be a binary blob or a parsed, structured metadata? If former, there is relatively little change involved in libcudf side, although not a pretty API. For latter, there needs to be translation between the struct we get and the format libcudf works with.
2. Does reading the metadata from cloud using `pyarrow.dataset` involve transferring anything more than just the footer blob over the network?
3. For typical parquet files in the wild, isn't the footer a very small size? I ask this because the sample file is 50,000 rows and still contains 10 row groups. The official [recommendation](https://github.com/apache/parquet-format#configurations) is 512MB per row-group so 50,000 might fit in one group, making for a very small footer.
4. @calebwin In the benchmarks you posted, is ~2ms just the IO or does it include parsing by pyarrow?

Reply:
> Will the data being passed to libcudf be a binary blob or a parsed, structured metadata? If former, there is relatively little change involved in libcudf side, although not a pretty API. For latter, there needs to be translation between the struct we get and the format libcudf works with.

The parsing of metadata happens at the arrow-cpp layer [1]. So either we would have to somehow use arrow-cpp directly or we would have to modify the pyarrow API to expose either re-serialization of metadata or access to the metadata blob and modify arrow-cpp to actually store the metadata blob instead of replacing it with the typed `FileMetaData` representation.

> Does reading the metadata from cloud using pyarrow.dataset involve transferring anything more than just the footer blob over the network?

I don't know. @jorisvandenbossche Do you have know how metadata reading works when reading from cloud? I know HDFS has a large block size but maybe when reading from the cloud, you only read in a common metadata file.

That could be one way to mitigate this issue - introduce a parameter in cuDF's `read_parquet` function to pass in a common metadata file in addition to the original files. For now, that file might just be passed directly to pyarrow but in the future, it could be used by libcudf directly. However, I do not actually know how to explicitly pass in a common metadata file to pyarrow. If we pass a directory to pyarrow, it will find the metadata file, but cuDF only accepts a set of files. To summarize, there would have to be some API change to cuDF and possibly also an additional API change to pyarrow to make use of common metadata files.

> For typical parquet files in the wild, isn't the footer a very small size? I ask this because the sample file is 50,000 rows and still contains 10 row groups. The official recommendation is 512MB per row-group so 50,000 might fit in one group, making for a very small footer.

Yes. I can do additional benchmarks if that would be helpful but my original thinking was also that given this is a very small dataset  (50,000 rows) with more row groups than needed (10 row groups), the metadata reading/parsing time would be exaggerated and in practice an overhead less than ~15% would be expected.

> @calebwin In the benchmarks you posted, is ~2ms just the IO or does it include parsing by pyarrow?

The time in the benchmarks includes I/O and parsing.

[1] https://github.com/apache/arrow/blob/2e6009621011d7df43882aa883905b84d1647018/cpp/src/parquet/file_reader.cc#L268

Reply:
> The time in the benchmarks includes I/O and parsing.

Would it be possible to get the breakup?

> Do you have know how metadata reading works when reading from cloud?

This would be helpful to know. I'm not versed in HDFS but if reading from it involves reading an entire block at a time then reusing just the metadata read might not be enough.

Reply:
> > The time in the benchmarks includes I/O and parsing.
> 
> Would it be possible to get the breakup?

Well, I would need to figure out how much time is spent on [this function call](https://github.com/apache/arrow/blob/2e6009621011d7df43882aa883905b84d1647018/cpp/src/parquet/file_reader.cc#L532) which parses metadata (which is encoded using the Thrift compact protocol). I'm not really sure how to do that - any pointers?

> > Do you have know how metadata reading works when reading from cloud?
>
> This would be helpful to know. I'm not versed in HDFS but if reading from it involves reading an entire block at a time > then reusing just the metadata read might not be enough.

Yeah, I'm not 100% sure but from what I have read - HDFS is "just" an abstraction over underlying operating system and so when pyarrow is invoked on each file to read a small blob of metadata, only that blob is read.

Reply:
Can this be marked resolved for this PR?

Further discussion should happen here - https://github.com/rapidsai/cudf/issues/6004.